src/mbgl/text/placement.cpp

```diff
@@ -0,0 +1,310 @@
+            if (symbolInstance.iconQuad) {
+                bucket.icon.opacityVertices.emplace_back(opacityVertex);
+                bucket.icon.opacityVertices.emplace_back(opacityVertex);
+                bucket.icon.opacityVertices.emplace_back(opacityVertex);
+                bucket.icon.opacityVertices.emplace_back(opacityVertex);
+            }
+            if (symbolInstance.placedIconIndex) {
+                bucket.icon.placedSymbols[*symbolInstance.placedIconIndex].hidden = opacityState.isHidden();
+            }
+        }
+        
+        auto updateCollisionBox = [&](const auto& feature, const bool placed) {
+            for (const CollisionBox& box : feature.boxes) {
+                if (feature.alongLine) {
+                   auto dynamicVertex = CollisionBoxDynamicAttributes::vertex(placed, !box.used);
+                    bucket.collisionCircle.dynamicVertices.emplace_back(dynamicVertex);
+                    bucket.collisionCircle.dynamicVertices.emplace_back(dynamicVertex);
+                    bucket.collisionCircle.dynamicVertices.emplace_back(dynamicVertex);
+                    bucket.collisionCircle.dynamicVertices.emplace_back(dynamicVertex);
+                } else {
+                    auto dynamicVertex = CollisionBoxDynamicAttributes::vertex(placed, false);
+                    bucket.collisionBox.dynamicVertices.emplace_back(dynamicVertex);
+                    bucket.collisionBox.dynamicVertices.emplace_back(dynamicVertex);
+                    bucket.collisionBox.dynamicVertices.emplace_back(dynamicVertex);
+                    bucket.collisionBox.dynamicVertices.emplace_back(dynamicVertex);
+                }
+            }
+        };
+        updateCollisionBox(symbolInstance.textCollisionFeature, opacityState.text.placed);
+        updateCollisionBox(symbolInstance.iconCollisionFeature, opacityState.icon.placed);
+    }
+
+    bucket.updateOpacity();
+    bucket.sortFeatures(state.getAngle());
+}
+
+JointOpacityState Placement::getOpacity(uint32_t crossTileSymbolID) const {
+    auto it = opacities.find(crossTileSymbolID);
+    if (it != opacities.end()) {
+        return it->second;
+    } else {
+        return JointOpacityState(false, false);
+    }
+
+}
+
+float Placement::symbolFadeChange(TimePoint now) const {
+    if (mapMode == MapMode::Continuous) {
+        return std::chrono::duration<float>(now - commitTime) / Duration(std::chrono::milliseconds(300));
+    } else {
+        return 1.0;
+    }
+}
+
+bool Placement::hasTransitions(TimePoint now) const {
+    return symbolFadeChange(now) < 1.0 || stale;
+}
+
+bool Placement::stillRecent(TimePoint now) const {
+    return mapMode == MapMode::Continuous && recentUntil > now;
+}
+void Placement::setRecent(TimePoint now) {
+    stale = false;
+    if (mapMode == MapMode::Continuous) {
+        // Only set in continuous mode because "now" isn't defined in still mode
+        recentUntil = now + Duration(std::chrono::milliseconds(300));
```

Comment:
Should we make this configurable?

Reply:
I'd lean against making placement frequency configurable with the idea that it's hard to reason about the effects of changing it.